tests/org.eclipse.swt.tests.gtk/JUnit Tests/org/eclipse/swt/tests/gtk/Test_Gtk_Tree_Virtual_setImage.java

```diff
@@ -0,0 +1,185 @@
+import org.eclipse.swt.graphics.Image;
+import org.eclipse.swt.layout.GridData;
+import org.eclipse.swt.layout.GridLayout;
+import org.eclipse.swt.widgets.Shell;
+import org.eclipse.swt.widgets.Tree;
+import org.eclipse.swt.widgets.TreeItem;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+public class Test_Gtk_Tree_Virtual_setImage {
+
+	private static final int WINDOW_WIDTH = 300;
+	private static final int WINDOW_HEIGHT = 100;
+	private static final int IMAGE_SIZE = 70;
+
+	private Shell shell;
+	private Tree tree;
+	private Image image;
+
+	@Before
+	public void setUp() {
+		shell = new Shell();
+		shell.setLayout(new GridLayout());
+	}
+
+	@After
+	public void tearDown() {
+		if (tree != null) {
+			tree.dispose();
+			tree = null;
+		}
+		if (image != null) {
+			image.dispose();
+			image = null;
+		}
+		if (shell != null) {
+			shell.dispose();
+			shell = null;
+		}
+	}
+
+	@Test
+	public void test_initial_img_5_2() {
+		do_test_initial_img(false, 5, 2);
+	}
+
+	@Test
+	public void test_initial_img_5_2_wCheckbox() {
+		do_test_initial_img(true, 5, 2);
+	}
+
+	@Test
+	public void test_initial_img_50_22() {
+		do_test_initial_img(false, 50, 22);
+	}
+
+	@Test
+	public void test_initial_img_50_22_wCheckbox() {
+		do_test_initial_img(true, 50, 22);
+	}
+
+	@Test
+	public void test_set_img_50_22() {
+		do_test_set_img(false, 50, 22);
+	}
+
+	@Test
+	public void test_set_img_50_22_wCheckbox() {
+		do_test_set_img(true, 50, 22);
+	}
+
+	private void do_test_initial_img(boolean withCheckbox, int totalRows, int imgRowIdx) {
+		createImage();
+		createTree(withCheckbox, totalRows, imgRowIdx, () -> image);
+		shell.pack();
+		shell.open();
+
+		executePendingUiTasks();
+
+		tree.showItem(tree.getItem(imgRowIdx));
+
+		executePendingUiTasks();
+
+		assertAllTreeRowsResized();
+	}
+
+	private void do_test_set_img(boolean withCheckbox, int totalRows, int imgRowIdx) {
+		createTree(withCheckbox, totalRows, imgRowIdx, () -> image);
+		shell.pack();
+		shell.open();
+
+		executePendingUiTasks();
+
+		tree.showItem(tree.getItem(imgRowIdx));
+
+		executePendingUiTasks();
+
+		tree.showItem(tree.getItem(totalRows - 1));
+
+		executePendingUiTasks();
+
+		createImage();
+
+		executePendingUiTasks();
+
+		tree.clear(imgRowIdx, false);
+
+		executePendingUiTasks();
+
+		tree.showItem(tree.getItem(imgRowIdx));
+
+		executePendingUiTasks();
+
+		assertAllTreeRowsResized();
+	}
+
+	private void createImage() {
+		this.image = new Image(shell.getDisplay(), IMAGE_SIZE, IMAGE_SIZE);
+		var color = shell.getDisplay().getSystemColor(SWT.COLOR_GREEN);
+		var gc = new GC(image);
+		gc.setForeground(color);
+		gc.setBackground(color);
+		gc.fillRectangle(0, 0, IMAGE_SIZE, IMAGE_SIZE);
+		gc.dispose();
+	}
+
+	private void createTree(boolean withCheckbox, int totalRows, int imgRowIdx, Supplier<Image> getImage) {
+		tree = new Tree(shell, SWT.VIRTUAL | SWT.BORDER | (withCheckbox ? SWT.CHECK : 0));
```

Comment:
[mino] Now, that a potential regression for setting an image of decreasing sizes in different columns is discovered, we may want some tests for multi-column case.

Reply:
It seems to me that `Tree` doesn't support different images sizes in swt by design.  
In [this reply](https://github.com/eclipse-platform/eclipse.platform.swt/pull/1662#discussion_r1886392687) I included a snippet that demonstrates what happens when there are multiple columns.  
Please, note that images are scaled to fixed size inside `ImageList`, which is pretty old - that's why I think that `Tree` expects images of the same size by design.

Reply:
BTW, I can add unit test(s) for multiple columns and different images sizes to `Test_Gtk_Tree_Virtual_setImage` no problem.  
I just don't know what additional cases you want to be checked.  
The current unit tests in `Test_Gtk_Tree_Virtual_setImage` are for the bug with row heights, that I noticed (and fixed in this PR) while I've been fixing crash in #678.

Reply:
The scenario is:
- for three sequences of images three in length each:
  - one monotonically increasing in size
  - one monotonically decreasing
  - and one non-monotonic
- test that behavior (heights, widths, exceptions, ideally cropping but that's probably too hard) is unchanged when:
  - adding images to different columns 
  - adding to same column
  - adding to a tree without columns

Reply:
OK.  
As I understand, you are working on some task related to image cropping, multiple image sizes, and multiple columns.  
I suppose including those unit tests in the commit for this task I would be more appropriate.

Reply:
No, I literally tried to fix the very same problem - crash in VIRTUAL Tree. While I succeeded, your solution might be a bit better.

Reply:
No, I [fixed](https://github.com/eclipse-platform/eclipse.platform.swt/pull/1612) the very same problem - crash in VIRTUAL Tree, but your solution might be better.